### PR TITLE
413 changehandle

### PR DESCRIPTION
**Pull request by claireolmstead**
# Problem

We do not have an endpoint to return an encoded payload for signing a change handle call.

Link to GitHub Issue(s): #413 

# Solution

- Create GET endpoint to return an encoded payload.
- Remove encoding from the POST change handle.

with @JoeCap08055 @enddynayn 

## Steps to Verify:

1. Get encoded extrinsic
2. Post change handle

### CODE REVIEW

File: services/account/apps/api/src/metadata.ts [hunk 1]
```diff
@@ -110,19 +114,23 @@ export default async () => {
...
         ],
         [
-          import('../../../libs/common/src/types/dtos/transaction.response.dto'),
-          { TransactionResponse: { referenceId: { required: true, type: () => String } } },
-        ],
-        [
-          import('../../../libs/common/src/types/dtos/wallet.login.response.dto'),
+          import('../../../libs/common/src/types/dtos/delegation.response.dto'),
           {
-            WalletLoginResponse: {
```
Comment (claireolmstead) [inline services/account/apps/api/src/metadata.ts:119]: I didn't change any of this code. was the metadata just out of date?

File: services/account/apps/api/src/controllers/v1/handles-v1.controller.ts [hunk 2]
```diff
@@ -69,6 +75,32 @@ export class HandlesControllerV1 {
     }
   }
 
+  @Get('change/:newHandle')
+  @HttpCode(HttpStatus.OK)
+  @ApiOperation({ summary: 'Get a properly encoded ClaimHandlePayload that can be signed.' })
+  @ApiOkResponse({ description: 'ClaimHandlePayload request enqueued' })
```
Comment (JoeCap08055) [inline services/account/apps/api/src/controllers/v1/handles-v1.controller.ts:81]:
```suggestion
  @ApiOkResponse({ description: 'Returned an encoded ClaimHandlePayload for signing' })
```

File: services/account/apps/api/src/services/handles.service.ts [hunk 3]
```diff
@@ -17,4 +19,13 @@ export class HandlesService {
     }
     throw new Error('Invalid msaId.');
   }
+
+  async getExpiration(): Promise<number> {
+    const lastFinalizedBlockNumber = await this.blockchainService.getLatestFinalizedBlockNumber();
+    return lastFinalizedBlockNumber + 10;
+  }
+
+  async encodePayload(payload: HandleRequest['payload']): Promise<CommonPrimitivesHandlesClaimHandlePayload> {
```
Comment (JoeCap08055) [inline services/account/apps/api/src/services/handles.service.ts:28]:
```suggestion
 encodePayload(payload: HandleRequest['payload']): CommonPrimitivesHandlesClaimHandlePayload {
```

File: services/account/apps/api/src/controllers/v1/handles-v1.controller.ts [hunk 4]
```diff
@@ -69,6 +75,32 @@ export class HandlesControllerV1 {
...
+   * Using the provided query parameters, creates a new payload that can be signed to change handle.
+   * @param queryParams - The query parameters for changing the handle.
+   * @returns Payload is included for convenience. Encoded payload to be used when signing the transaction.
+   * @throws An error if the change handle payload creation fails.
+   */
+  async getChangeHandlePayload(@Param('newHandle') newHandle: string): Promise<ChangeHandlePayloadRequest> {
+    try {
+      const expiration = await this.handlesService.getExpiration();
+      const payload = { baseHandle: newHandle, expiration };
+      const encodedPayload = await this.handlesService.encodePayload(payload);
```
Comment (JoeCap08055) [inline services/account/apps/api/src/controllers/v1/handles-v1.controller.ts:92]: So, this gives us a SCALE-encoded payload. But we need to sign raw bytes. So what we actually need to return is a Byte-wrapped array of bytes, encoded as a hex string. Like so:
```suggestion
      const encodedPayload = u8aToHex(u8aWrapBytes(this.handlesService.encodePayload(payload).toU8a()));
```

Also, `encodePayload` shouldn't be `async`, so we don't need `await`.

File: services/account/apps/api/src/services/handles.service.ts [hunk 5]
```diff
@@ -17,4 +19,13 @@ export class HandlesService {
     }
     throw new Error('Invalid msaId.');
   }
+
+  async getExpiration(): Promise<number> {
+    const lastFinalizedBlockNumber = await this.blockchainService.getLatestFinalizedBlockNumber();
+    return lastFinalizedBlockNumber + 10;
```
Comment (JoeCap08055) [inline services/account/apps/api/src/services/handles.service.ts:25]: FYI the standard expiration we adopted in SIWF for claiming a handle is 10 minutes (= 600 / SECONDS_PER_BLOCK) blocks.

Though I wonder if (time in seconds) should be an (optional) input parameter to the `GET` endpoint; if not provided we could assume 10 minutes. I think we should add that; it's easy enough. If there are no objections @mattheworris @claireolmstead , I'll add it to the issue AC.

File: services/account/libs/common/src/types/dtos/handles.request.dto.ts [hunk 6]
```diff
@@ -28,6 +30,16 @@ export class HandleRequest {
   proof: HexString;
 }
 
+export class ChangeHandlePayloadRequest {
+  @ApiProperty()
+  @IsNotEmpty()
```
Comment (JoeCap08055) [inline services/account/libs/common/src/types/dtos/handles.request.dto.ts:35]: FYI I don't think these `class-validator` decorators have any effect on return types; only on input types.

File: services/account/swagger.json
Comment (JoeCap08055) [inline services/account/swagger.json:1]: Don't forget to re-generate this file again after making more changes to the controller.

**Review by JoeCap08055 — CHANGES_REQUESTED** [review]

**Review by JoeCap08055 — APPROVED** [review]

File: services/account/apps/api/src/services/handles.service.ts [hunk 7]
```diff
@@ -17,4 +17,13 @@ export class HandlesService {
...
```
Comment (mattheworris) [inline services/account/apps/api/src/services/handles.service.ts:23]: 10 blocks may not be long enough to enable signing of this transaction

100@ 6 seconds or 50 @ 12 seconds should probably be a default constant from the config service
```suggestion
    return lastFinalizedBlockNumber + 100;
```
or if we add `@Query('expirationTime') expirationTime?: number,` to the endpoint.
```suggestion
    return lastFinalizedBlockNumber + expirationTime/SECS_PER_BLOCK; //Pseudocode
```

**Review by mattheworris — CHANGES_REQUESTED** [review]
- Read through changes
- Requesting updates to block expiration

**Review by mattheworris — APPROVED** [review]

**Pull request merged**